token.go

```diff
@@ -0,0 +1,315 @@
+		fallthrough
+	case 11:
+		k2 ^= uint64(tail[10]) << 16
+		fallthrough
+	case 10:
+		k2 ^= uint64(tail[9]) << 8
+		fallthrough
+	case 9:
+		k2 ^= uint64(tail[8])
+
+		k2 *= c2
+		k2 = (k2 << 33) | (k2 >> 31) // ROTL64(k2, 33)
+		k2 *= c1
+		h2 ^= k2
+
+		fallthrough
+	case 8:
+		k1 ^= uint64(tail[7]) << 56
+		fallthrough
+	case 7:
+		k1 ^= uint64(tail[6]) << 48
+		fallthrough
+	case 6:
+		k1 ^= uint64(tail[5]) << 40
+		fallthrough
+	case 5:
+		k1 ^= uint64(tail[4]) << 32
+		fallthrough
+	case 4:
+		k1 ^= uint64(tail[3]) << 24
+		fallthrough
+	case 3:
+		k1 ^= uint64(tail[2]) << 16
+		fallthrough
+	case 2:
+		k1 ^= uint64(tail[1]) << 8
+		fallthrough
+	case 1:
+		k1 ^= uint64(tail[0])
+
+		k1 *= c1
+		k1 = (k1 << 31) | (k1 >> 33) // ROTL64(k1, 31)
+		k1 *= c2
+		h1 ^= k1
+	}
+
+	h1 ^= uint64(length)
+	h2 ^= uint64(length)
+
+	h1 += h2
+	h2 += h1
+
+	// finalizer
+	const fmix1 = 0xff51afd7ed558ccd
+	const fmix2 = 0xc4ceb9fe1a85ec53
+
+	// fmix64(h1)
+	h1 ^= h1 >> 33
+	h1 *= fmix1
+	h1 ^= h1 >> 33
+	h1 *= fmix2
+	h1 ^= h1 >> 33
+
+	// fmix64(h2)
+	h2 ^= h2 >> 33
+	h2 *= fmix1
+	h2 ^= h2 >> 33
+	h2 *= fmix2
+	h2 ^= h2 >> 33
+
+	h1 += h2
+	// the following is extraneous since h2 is discarded
+	// h2 += h1
+
+	return h1
+}
+
+func (p Murmur3Partitioner) ParseString(str string) Token {
+	val, _ := strconv.ParseInt(str, 10, 64)
+	return Murmur3Token(val)
+}
+
+func (m Murmur3Token) String() string {
+	return strconv.FormatInt(int64(m), 10)
+}
+
+func (m Murmur3Token) Less(token Token) bool {
+	return m < token.(Murmur3Token)
+}
+
+// order preserving partitioner and token
+type OrderPreservingPartitioner struct{}
+type OrderPreservingToken []byte
+
+func (p OrderPreservingPartitioner) Hash(partitionKey []byte) Token {
+	// the partition key is the token
+	return OrderPreservingToken(partitionKey)
+}
+
+func (p OrderPreservingPartitioner) ParseString(str string) Token {
+	return OrderPreservingToken([]byte(str))
+}
+
+func (o OrderPreservingToken) String() string {
+	return string([]byte(o))
+}
+
+func (o OrderPreservingToken) Less(token Token) bool {
+	return -1 == bytes.Compare(o, token.(OrderPreservingToken))
+}
+
+// random partitioner and token
+type RandomPartitioner struct{}
+type RandomToken struct {
+	*big.Int
+}
+
+func (p RandomPartitioner) Hash(partitionKey []byte) Token {
+	hash := md5.New()
+	sum := hash.Sum(partitionKey)
+
+	val := new(big.Int)
+	val = val.SetBytes(sum)
+	val = val.Abs(val)
+
+	return RandomToken{val}
+}
+
+func (p RandomPartitioner) ParseString(str string) Token {
+	val := new(big.Int)
+	val.SetString(str, 10)
+	return RandomToken{val}
+}
+
+func (r RandomToken) Less(token Token) bool {
+	return -1 == r.Int.Cmp(token.(RandomToken).Int)
+}
+
+// a data structure for organizing the relationship between tokens and hosts
+type TokenRing struct {
+	partitioner Partitioner
+	tokens      []Token
+	hosts       []*HostInfo
+}
+
+func NewTokenRing(partitioner string, hosts []*HostInfo) (*TokenRing, error) {
```

Comment:
It'll be owned by the TokenAwareHostSelectionPolicy and when the policy get's updates from the connection pool when the host info changes, then it will replace its token ring.